syntax = "proto3";

package auth;

// ============ СЕРВИС ============

service AuthService {
    // Аутентификация
    rpc Register(RegisterRequest) returns (RegisterResponse);
    rpc Login(LoginRequest) returns (LoginResponse);
    rpc Logout(LogoutRequest) returns (LogoutResponse);
    rpc RefreshToken(RefreshTokenRequest) returns (RefreshTokenResponse);
    
    // Валидация
    rpc ValidateToken(ValidateTokenRequest) returns (ValidateTokenResponse);
    rpc GetCurrentUser(GetCurrentUserRequest) returns (UserResponse);
    
    // Управление пользователями
    rpc GetUser(GetUserRequest) returns (UserResponse);
    rpc UpdateUser(UpdateUserRequest) returns (UserResponse);
    rpc DeleteUser(DeleteUserRequest) returns (DeleteResponse);
    rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
    
    // Пароли
    rpc ChangePassword(ChangePasswordRequest) returns (ChangePasswordResponse);
    rpc RequestPasswordReset(PasswordResetRequest) returns (PasswordResetResponse);
    rpc ResetPassword(ResetPasswordRequest) returns (ResetPasswordResponse);
    
    // Роли
    rpc AssignRole(AssignRoleRequest) returns (AssignRoleResponse);
    rpc RevokeRole(RevokeRoleRequest) returns (RevokeRoleResponse);
    rpc GetUserRoles(GetUserRolesRequest) returns (UserRolesResponse);
    
    // Статистика
    rpc GetLoginHistory(LoginHistoryRequest) returns (LoginHistoryResponse);
    
    // Здоровье
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// ============ МОДЕЛИ ============

message User {
    int32 id = 1;
    string username = 2;
    string email = 3;
    string full_name = 4;
    string phone = 5;
    
    string primary_role = 6;
    repeated string roles = 7;
    
    int32 teacher_id = 8;
    int32 staff_id = 9;
    int32 student_group_id = 10;
    
    bool is_active = 11;
    bool is_verified = 12;
    
    string last_login_at = 13;
    string created_at = 14;
    string updated_at = 15;
}

message Role {
    int32 id = 1;
    string name = 2;
    string code = 3;
    string description = 4;
    map<string, bool> permissions = 5;
}

message TokenPair {
    string access_token = 1;
    string refresh_token = 2;
    int32 expires_in = 3;
    string token_type = 4;
}

// ============ АУТЕНТИФИКАЦИЯ ============

message RegisterRequest {
    string username = 1;
    string email = 2;
    string password = 3;
    string full_name = 4;
    string phone = 5;
    string primary_role = 6;
    
    int32 teacher_id = 7;
    int32 student_group_id = 8;
}

message RegisterResponse {
    bool success = 1;
    User user = 2;
    TokenPair tokens = 3;
    string message = 4;
}

message LoginRequest {
    string username = 1;
    string password = 2;
    
    string ip_address = 3;
    string user_agent = 4;
    string device_id = 5;
}

message LoginResponse {
    bool success = 1;
    User user = 2;
    TokenPair tokens = 3;
    string message = 4;
}

message LogoutRequest {
    string refresh_token = 1;
    int32 user_id = 2;
}

message LogoutResponse {
    bool success = 1;
    string message = 2;
}

message RefreshTokenRequest {
    string refresh_token = 1;
    string ip_address = 2;
    string user_agent = 3;
}

message RefreshTokenResponse {
    bool success = 1;
    TokenPair tokens = 2;
    string message = 3;
}

// ============ ВАЛИДАЦИЯ ============

message ValidateTokenRequest {
    string access_token = 1;
}

message ValidateTokenResponse {
    bool valid = 1;
    int32 user_id = 2;
    string username = 3;
    string primary_role = 4;
    repeated string roles = 5;
    int32 expires_at = 6;
    string message = 7;
}

message GetCurrentUserRequest {
    string access_token = 1;
}

// ============ CRUD ПОЛЬЗОВАТЕЛЕЙ ============

message GetUserRequest {
    oneof identifier {
        int32 id = 1;
        string username = 2;
        string email = 3;
    }
}

message UserResponse {
    User user = 1;
    string message = 2;
}

message UpdateUserRequest {
    int32 id = 1;
    
    string full_name = 2;
    string email = 3;
    string phone = 4;
    bool is_active = 5;
    
    int32 updated_by = 6;
}

message DeleteUserRequest {
    int32 id = 1;
    bool hard_delete = 2;
}

message DeleteResponse {
    bool success = 1;
    string message = 2;
}

message ListUsersRequest {
    int32 page = 1;
    int32 page_size = 2;
    
    repeated string roles = 3;
    bool only_active = 4;
    string search_query = 5;
    
    string sort_by = 6;
    string sort_order = 7;
}

message ListUsersResponse {
    repeated User users = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

// ============ ПАРОЛИ ============

message ChangePasswordRequest {
    int32 user_id = 1;
    string old_password = 2;
    string new_password = 3;
}

message ChangePasswordResponse {
    bool success = 1;
    string message = 2;
}

message PasswordResetRequest {
    string email = 1;
    string ip_address = 2;
}

message PasswordResetResponse {
    bool success = 1;
    string message = 2;
}

message ResetPasswordRequest {
    string token = 1;
    string new_password = 2;
}

message ResetPasswordResponse {
    bool success = 1;
    string message = 2;
}

// ============ РОЛИ ============

message AssignRoleRequest {
    int32 user_id = 1;
    string role_code = 2;
    int32 granted_by = 3;
}

message AssignRoleResponse {
    bool success = 1;
    string message = 2;
}

message RevokeRoleRequest {
    int32 user_id = 1;
    string role_code = 2;
}

message RevokeRoleResponse {
    bool success = 1;
    string message = 2;
}

message GetUserRolesRequest {
    int32 user_id = 1;
}

message UserRolesResponse {
    repeated Role roles = 1;
}

// ============ ИСТОРИЯ ============

message LoginHistoryRequest {
    int32 user_id = 1;
    int32 limit = 2;
    bool only_successful = 3;
}

message LoginHistoryResponse {
    repeated LoginHistoryEntry entries = 1;
}

message LoginHistoryEntry {
    int32 id = 1;
    string username = 2;
    bool success = 3;
    string ip_address = 4;
    string user_agent = 5;
    string failure_reason = 6;
    string created_at = 7;
}

message HealthCheckRequest {}

message HealthCheckResponse {
    string status = 1;
    string version = 2;
}

